### PR TITLE
[RL]update big run config

### PR DESCRIPTION
**Pull request by yichuan-w**
## Summary

Makes the **Qwen3-30B-A3B DeepEP Search-R1** big run actually train **and learn**
(held-out validation reward rising), by baking the working configuration into the
recipe `rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf` plus the few correctness/observability
fixes it needs. On MAST (4 generators, FSDP=16/EP=16 trainer over 2 hosts) this run goes
from cold-start to learning: **validation_reward 0.25 (step 0) -> 0.39 (step 20)**, with
**0** zero-gradient steps (vs ~33% before the drop_zero_std fix).

## Recipe knobs (`rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf`)

| Knob | Value | Why |
|------|-------|-----|
| `drop_zero_std_reward_groups` | `True` | **The decisive learning fix.** `False` kept all-same-reward (zero-advantage) groups: ~1/3 of steps had `grad_norm==0` and the rest were diluted, so validation reward stayed flat (~0.20-0.25 for 100+ steps). FULL_DECODE_ONLY gives ~25% nonzero rollouts from cold start, so `True` is safe (steps still land) and drops the dead groups for a full gradient -> validation rises. |
| generator `cudagraph` | `FULL_DECODE_ONLY` | The DeepEP EXPAND `FULL_AND_PIECEWISE` cudagraph corrupted the MoE forward (garbage output, never searched, 0 reward). Decode-only graphing (prefill/mixed eager) avoids it -- the same #3668 fix the 1.7B/8B/dense configs use -- and is ~10x faster than full eager (decode ~4500ms vs ~50000ms). |
| generator `override` | `helion_rope` + `deepep_inference` (NOT `fused_swiglu`) | `fused_swiglu`'s fused_grouped_experts claims `moe.experts` while `deepep_inference` claims the descendant `moe.experts.token_dispatcher` -- a nested-override conflict (#3778). The trainer keeps `fused_swiglu`. |
| trainer `loss` | `ChunkedLossWrapper(num_chunks=16, DAPOLoss)` | A bare DAPOLoss materializes the full-vocab (151936) x seq logprob transient in one shot and OOMs the 80GB trainer. |
| trainer `parallelism` | FSDP=16 / EP=16 (two hosts) | FSDP=8/EP=8 sits at the 80GB edge and a step-2 fwd/bwd tripped a CUDA "unspecified launch failure"; 16-way roughly halves per-GPU memory. |
| `local_batch_size` | 2 | `local_batch_size=1` split each step into many microbatches, each paying the full inter-node EP all-to-all latency -> low trainer MFU. Packing 2 rows/forward roughly halves the microbatch count (~1.4-1.7x faster fwd/bwd). |
| `group_size` / `num_groups_per_train_step` | 8 / 8 | 64 rollouts/step; works with `drop_zero_std=True` since FULL_DECODE_ONLY keeps most groups non-zero-std. |
| `num_generators` | 4 | Generation is the async-loop bottleneck; the GeneratorRouter load-balances rollouts. The single controller retrieval server caps how far this scales. |
| `validation.interval` | 20 | Periodic held-out validation so the validation_reward curve is readable (per-step train reward is too noisy); runs on the otherwise-idle generators. |

Mixed precision is unchanged (TrainingConfig default: fp32 master + bf16 compute + fp32
reduce, which DeepEP requires); the rubric is unchanged (pure exact-match).

## Supporting fixes

- **Weight sync keeps fp32 buffers** (`actors/trainer.py`): builds on #3825 (exclude
  buffers from the fp32->bf16 cast so the MoE `expert_bias_E` stays fp32) and adds FQN
  normalization -- activation checkpointing injects `._checkpoint_wrapped_module` and
  `torch.compile` injects `._orig_mod` into `named_buffers()` names but `state_dict()`
  strips them, so a wrapped buffer was misclassified as a param and wrongly cast. The
  30B recipe uses SelectiveAC, which hit exactly this.
- **Periodic held-out validation** (`controller.py`): `ValidationConfig.interval` runs a
  validation pass every `interval` optimizer steps (0 = pre/post only, the old behavior).
- **Validation reward console key** (`observability/metrics/processor.py`): log
  `validation_reward/*` so the held-out reward (accuracy under pure EM) reaches the
  metrics sink.

## Test plan / validation

Qwen3-30B-A3B, DeepEP, Search-R1 on MAST (trainer FSDP=16/EP=16 over two hosts, four
generators TP=4/EP=4, faiss-GPU retriever on the controller):

- **Weight sync:** trainer keeps all `expert_bias_E` buffers fp32; sync succeeds.
- **Generation:** FULL_DECODE_ONLY produces correct terse answers (validation
  response_length ~13.6, the model searches), not the ~505-char garbage of the corrupt
  cudagraph path.
- **Learning:** with `drop_zero_std=True`, **validation_reward rises 0.25 -> 0.39** over
  the first 20 steps with **0** `grad_norm==0` steps; before the fix it was flat at
  ~0.20-0.25 with ~33% dead steps.

## Notes

- `FULL_DECODE_ONLY` (vs `FULL_AND_PIECEWISE`) and the eager-vs-graph history are a
  workaround for a real DeepEP EXPAND cudagraph bug; fixing the EXPAND capture and
  restoring full cudagraphs is a follow-up.
- `drop_zero_std=True` relies on the policy scoring some nonzero reward at cold start
  (FULL_DECODE_ONLY ensures ~25%); a recipe whose cold start is all-zero would need the
  `False` bootstrap first.

### CODE REVIEW

**Comment by yichuan-w** [comment]
Opening as draft for visibility in case anyone hits the same crash.

**One-line:** for MoE policies, the fp32-trainer -> bf16-generator weight sync was
casting the router's fp32 `expert_bias_E` buffer to bf16, which trips the generator's
get-side dtype check (`bfloat16 != float32`) on the first sync. The fix casts only
parameters and leaves registered buffers at their own dtype.

Verified on Qwen3-30B-A3B (DeepEP, Search-R1): the crash is gone and the run gets past
the first weight sync. Keeping it draft until I've self-reviewed -- feedback welcome.

File: torchtitan/experiments/rl/actors/trainer.py [hunk 1]
```diff
@@ -496,23 +496,32 @@ async def push_model_state_dict(self) -> None:
...
-            # cast to the generator dtype here (else the generator reads fp32 into its bf16 state dict).
-            # Exclude buffers from the cast: FSDP mixed precision casts params to the compute dtype but
-            # leaves buffers at their registered dtype (same as pretraining), e.g. the fp32
-            # expert_bias_E load-balance bias in MoE. The generator keeps those buffers at the same
-            # registered dtype, so casting them here would mismatch its state dict and fail torchstore's
-            # dtype check on weight sync.
-            # TODO(async-rl): remove this manual cast once torchstore applies transfer_dtype on the
-            #   CPU-staged path.
-            buffer_names = {name for name, _ in self.model.named_buffers()}
+            # torchstore only applies `transfer_dtype` on the RDMA path, so under
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: please rebase to main for "proper fix"

File: torchtitan/experiments/rl/controller.py [hunk 2]
```diff
@@ -1094,5 +1100,15 @@ async def _trainer_loop(
                     step, last_step=(step == num_training_steps)
                 )
 
+            # Periodic held-out validation. Finish the in-flight weight sync first so the
+            # generators hold this step's policy, then run a validation pass on the shared
+            # engine (disjoint negative request ids keep it separate from training
+            # rollouts). The final step is covered by run()'s post-training pass.
+            val_interval = self.config.async_loop.validation.interval
+            if val_interval and step % val_interval == 0 and step != num_training_steps:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/controller.py:1115]: what if `val_interval` is 0?

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 3]
```diff
@@ -254,3 +293,83 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
             continue
         moe.experts.token_dispatcher.num_max_tokens_per_rank = num_max_tokens_per_rank
     return config
+
+
+def rl_grpo_qwen3_30b_a3b_dense_search_r1_perf() -> Controller.Config:
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: could you directly update https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/examples/search_r1/config_registry.py#L155

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 4]
```diff
@@ -254,3 +293,83 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
+            model_dtype="bfloat16",
+            parallelism=InferenceParallelismConfig(  # single node generator, no EP
+                data_parallel_degree=1,
+                tensor_parallel_degree=4,
+            ),
+            # FULL_DECODE_ONLY (default): graph only pure-decode batches; run prefill /
+            # mixed batches eager. The known-good 1.7B/8B search_r1 configs use this;
+            # FULL_AND_PIECEWISE (also graphs prefill/mixed) corrupted generation here
+            # (gibberish output) -- see #3668. Match the small configs to derisk.
+            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_DECODE_ONLY"),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: let's try
```suggestion
            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_AND_PIECEWISE"),
```

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 5]
```diff
@@ -254,3 +293,83 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
+                keep_latest_k=3,
+            ),
+            loss=DAPOLoss.Config(ratio_clip_low=0.2, ratio_clip_high=0.28),
+            override=OverrideConfig(imports=list(perf_imports)),
+        ),
+        generator=VLLMGenerator.Config(
+            model_dtype="bfloat16",
+            parallelism=InferenceParallelismConfig(  # single node generator, no EP
+                data_parallel_degree=1,
+                tensor_parallel_degree=4,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: where's EP?

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 6]
```diff
@@ -254,3 +293,83 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
+        "torchtitan.overrides.helion_rope",
+    ]
+
+    return Controller.Config(
+        model_spec=model_spec,
+        hf_assets_path="torchtitan/experiments/rl/example_checkpoint/Qwen3-30B-A3B",
+        num_generators=2,  # TODO: TBD -- number of generator proc meshes to spawn
+        async_loop=AsyncLoopConfig(
+            num_training_steps=500,
+            num_groups_per_train_step=32,  # TODO: TBD
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py]: you said 8?

File: torchtitan/experiments/rl/actors/trainer.py [hunk 7]
```diff
@@ -496,23 +496,32 @@ async def push_model_state_dict(self) -> None:
...
+            # torchstore's get-side dtype check. named_buffers() FQNs carry module-wrapper
+            # segments (activation checkpoint's `._checkpoint_wrapped_module`, compile's
+            # `._orig_mod`) that state_dict() strips, so normalize both before matching,
+            # else a wrapped buffer is misclassified as a param and wrongly cast.
+            # TODO(async-rl): remove this manual cast once torchstore applies transfer_dtype
+            #   on the CPU-staged path.
+            def _clean(n: str) -> str:
+                return n.replace("._checkpoint_wrapped_module", "").replace(
+                    "._orig_mod", ""
+                )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: This seems related to AC. @wwwjn  Have you tested your fix works with SAC?

Comment (wwwjn) [reply torchtitan/experiments/rl/actors/trainer.py]: I tested locally with debug moe model, might be ignored incidentially. This change seems to be needed. Let me double check

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 8]
```diff
@@ -221,16 +256,20 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
+            # avoids it (#3668, as the 1.7B/8B/dense configs) and is far faster than eager.
+            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_DECODE_ONLY"),
             checkpoint=CheckpointManager.Config(enable=False),
             sampling=SamplingConfig(temperature=1.0, top_p=1.0, max_tokens=512),
-            # Generator-only: the DeepEP cudagraph EXPAND override on top of the perf
-            # overrides; the trainer keeps the compact path.
+            # Generator override: helion_rope + the DeepEP EXPAND inference override, but
+            # NOT fused_swiglu -- its fused_grouped_experts claims `moe.experts` while
+            # deepep_inference claims the descendant `moe.experts.token_dispatcher`, a
+            # nested-override conflict (#3778). The trainer keeps fused_swiglu.
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:269]: oh ... this is indeed a conflict. Need to note and think about it.

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 9]
```diff
@@ -199,10 +224,14 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
             lr_scheduler=LRSchedulersContainer.Config(
                 warmup_steps=2, decay_type="linear", min_lr_factor=1.0
             ),
+            # FSDP=16/EP=16 (two H100 hosts). FSDP=32/EP=8 over four hosts kept the EP
+            # all-to-all intra-node but moved the FSDP all-gather inter-node across four
+            # nodes -- net no speedup and 2 extra hosts, so reverted. FSDP=8/EP=8 on one
+            # host OOM'd, hence 16.
             parallelism=ParallelismConfig(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:231]: could you apply spmd_types backend for both trainer and generator?

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 10]
```diff
@@ -221,16 +256,20 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
                 expert_parallel_degree=4,
             ),
-            # varlen attention -> FULL_AND_PIECEWISE (decode captured FULL, prefill
-            # piecewise).
-            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_AND_PIECEWISE"),
+            # FULL_DECODE_ONLY: graph pure-decode batches, run prefill / mixed eager.
+            # FULL_AND_PIECEWISE (graphs prefill / mixed too) corrupted the DeepEP EXPAND
+            # MoE forward -> garbage output, no search, 0 reward. Decode-only graphing
+            # avoids it (#3668, as the 1.7B/8B/dense configs) and is far faster than eager.
+            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_DECODE_ONLY"),
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:263]:
```suggestion
            cudagraph=VLLMCudagraphConfig(enable=True, mode="FULL_AND_PIECEWISE"),
```

File: torchtitan/experiments/rl/actors/trainer.py (hunk 7 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/actors/trainer.py]: https://github.com/pytorch/torchtitan/pull/3861 I find a existing helper function to solve this

File: torchtitan/experiments/rl/observability/metrics/processor.py [hunk 11]
```diff
@@ -79,8 +79,11 @@ class Config(Configurable.Config):
 
         console_log_keys_validation: list[str] | None = field(
             default_factory=lambda: [
-                "validation/reward/_mean",
-                "validation/reward/_max",
+                # compute_rollout_metrics emits the reward under "<prefix>_reward" (an
+                # UNDERSCORE), matching the train side's "rollout_reward/_mean" -- so the
+                # validation reward key is "validation_reward/...", not "validation/reward/...".
```
Comment (felipemello1) [inline torchtitan/experiments/rl/observability/metrics/processor.py:84]:
```suggestion
```

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 12]
```diff
@@ -177,15 +181,36 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
     config = Controller.Config(
         model_spec=model_spec,
         hf_assets_path="torchtitan/experiments/rl/example_checkpoint/Qwen3-30B-A3B",
-        num_generators=2,  # TODO: TBD -- number of generator proc meshes to spawn
+        # Generation is the async-loop bottleneck; run several generators (the
```
Comment (felipemello1) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:184]: all the comments here are very verbose. Can we shorten things a bit?

File: torchtitan/experiments/rl/controller.py (hunk 2 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/controller.py:1115]: it would fail "if val_interval" right?

File: torchtitan/experiments/rl/controller.py [hunk 13]
```diff
@@ -1101,5 +1107,15 @@ async def _trainer_loop(
                     step, last_step=(step == num_training_steps)
                 )
 
+            # Periodic held-out validation. Finish the in-flight weight sync first so the
+            # generators hold this step's policy, then run a validation pass on the shared
+            # engine (disjoint negative request ids keep it separate from training
+            # rollouts). The final step is covered by run()'s post-training pass.
```
Comment (felipemello1) [inline torchtitan/experiments/rl/controller.py:1113]: lets make it 1-2 lines max, and no need to say "disjoint negative". Just mention that validation dont make it into the buffer.

**Review by felipemello1 — CHANGES_REQUESTED** [review]
some minor changes, but looks good!

File: torchtitan/experiments/rl/examples/search_r1/config_registry.py [hunk 14]
```diff
@@ -221,16 +256,20 @@ def rl_grpo_qwen3_30b_a3b_deepep_search_r1_perf() -> Controller.Config:
...
-            # Generator-only: the DeepEP cudagraph EXPAND override on top of the perf
-            # overrides; the trainer keeps the compact path.
+            # Generator override: helion_rope + the DeepEP EXPAND inference override, but
+            # NOT fused_swiglu -- its fused_grouped_experts claims `moe.experts` while
+            # deepep_inference claims the descendant `moe.experts.token_dispatcher`, a
+            # nested-override conflict (#3778). The trainer keeps fused_swiglu.
             override=OverrideConfig(
                 imports=[
-                    *perf_imports,
+                    "torchtitan.overrides.helion_rope",
```
Comment (wwwjn) [inline torchtitan/experiments/rl/examples/search_r1/config_registry.py:272]: This need to be rebased on to https://github.com/pytorch/torchtitan/pull/3894
